### PR TITLE
V2pipeline

### PR DESCRIPTION
**Pull request by ywrt**
Refactor the training input pipeline.

1. Introduce a binary format (v2).
2. Use the v2 format for the shuffle buffer (shrinks RAM by ~2x)
3. Support for loading v2 format off disk (much faster to parse)
4. Amend test to be full end-to-end test, including worker interfaces.
5. Add benchmark for the tensorflow frontend.
6. Change worker->main process communication to be fixed size binary format instead of protobuf.

Net result is that the pipeline now saturates at about 45,000 moves/sec into tensorflow (aka ~88 batches/sec with n=512).

### CODE REVIEW

**Comment by ywrt** [comment]
I got frustrated by the parsing speed when I was trying to run some experiments, so I changed a few things.

I think the only feature that it useful with v1 format training files is the larger shuffle buffer, albeit with the tradeoff that the buffer is split over the workers.
But if training files are available the (binary) v2 format, then the shuffle buffer loads at ~ 80,000 moves/second (aka ~6 seconds to fill).

**Comment by Alderi-Tokori** [comment]
@ywrt 45'000 positions per second? So a x10 speedup in training? Or have I misunderstood something? If that speeds training up that much it's incredible.

**Comment by ywrt** [comment]
No, the training itself is still GPU limited. This just reduces up the amount of CPU that's spend on data ingestion.

My motivation was primarily because I wanted to play around with training drastically simpler networks, where the training time was dominated by the data ingestion rate. It just happens that this code also reduces the memory footprint of the shuffle buffer et al

File: training/tf/parse.py [hunk 1]
```diff
@@ -52,7 +62,36 @@ def remap_vertex(vertex, symmetry):
...
+            The data is represented in a number of formats through this dataflow
+            pipeline. In order, they are:
+
+            chunk: The name of a file containing chunkdata
+
+            chunkdata: type Bytes. Either mutiple records of v1 format, or multiple records
+            of v2 format.
+
+            v1: The original text format describing a move. 19 lines long. VERY slow
+            to decode. Typically around 2500 bytes long. Used only for backward
```
Comment (sethtroisi) [inline training/tf/parse.py]: should this be 27568 (4 + 4 * 393 + 4 * 6498)?

**Comment by zediir** [comment]
Looking at the code this should still be backwards compatible with the v1 chuck files? I'm getting the following error:

```text
D:\My Folders\Documents\leela-zero-ywrt\models>..\training\tf\parse.py ..\..\leela-zero\traindatatygem\train.out
Found 18350 chunks
Training with 16516 chunks, validating on 1834 chunks
Using 6 worker processes.
Traceback (most recent call last):
  File "D:\My Folders\Documents\leela-zero-ywrt\training\tf\parse.py", line 535, in <module>
    main(sys.argv[1:])
  File "D:\My Folders\Documents\leela-zero-ywrt\training\tf\parse.py", line 507, in main
    shuffle_size=1<<19, sample=DOWN_SAMPLE)
  File "D:\My Folders\Documents\leela-zero-ywrt\training\tf\parse.py", line 138, in __init__
    args=(chunkdatagen, write)).start()
  File "C:\Users\zedii\AppData\Local\Programs\Python\Python36\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Users\zedii\AppData\Local\Programs\Python\Python36\lib\multiprocessing\context.py", line 223, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Users\zedii\AppData\Local\Programs\Python\Python36\lib\multiprocessing\context.py", line 322, in _Popen
    return Popen(process_obj)
  File "C:\Users\zedii\AppData\Local\Programs\Python\Python36\lib\multiprocessing\popen_spawn_win32.py", line 65, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\zedii\AppData\Local\Programs\Python\Python36\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: can't pickle Struct objects

D:\My Folders\Documents\leela-zero-ywrt\models>Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\zedii\AppData\Local\Programs\Python\Python36\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\zedii\AppData\Local\Programs\Python\Python36\lib\multiprocessing\spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
EOFError: Ran out of input
```

File: training/tf/parse.py [hunk 2]
```diff
@@ -104,76 +165,195 @@ def convert_train_data(self, text_item, symmetry):
...
+        probabilities = np.array(text_item[17].split()).astype(np.float32)
+        if np.any(np.isnan(probabilities)):
+            # Work around a bug in leela-zero v0.3, skipping any
+            # positions that have a NaN in the probabilities list.
+            return False, None
+        assert len(probabilities) == 362
+
+        probs = probabilities.tobytes()
+        assert(len(probs) == 362 * 4)
+
```
Comment (sethtroisi) [inline training/tf/parse.py]: nit -nl

File: training/tf/parse.py [hunk 3]
```diff
@@ -104,76 +165,195 @@ def convert_train_data(self, text_item, symmetry):
 
         # We flatten to a single array of len 16*19*19, type=np.uint8
         planes = np.concatenate(planes)
+        # and then to a byte string
+        planes = np.packbits(planes).tobytes()
 
+        # Get the 'side to move'
+        stm = text_item[16][0]
+        assert stm == "0" or stm == "1"
```
Comment (sethtroisi) [inline training/tf/parse.py]: consider: `assert stm in ("0", "1"), stm`

File: training/tf/parse.py [hunk 4]
```diff
@@ -104,76 +165,195 @@ def convert_train_data(self, text_item, symmetry):
...
+    def convert_chunkdata_to_v2(self, chunkdata):
+        """
+            Take chunk of unknown format, and return it as a list of
+            v2 format records.
+        """
+        if chunkdata[0:4] == b'\1\0\0\0':
+            #print("V2 chunkdata")
+            items = [ chunkdata[i:i+self.v2_struct.size]
+                        for i in range(0, len(chunkdata), self.v2_struct.size) ]
+            if self.sample > 0:
```
Comment (sethtroisi) [inline training/tf/parse.py]: `self.sample > 1`

File: training/tf/parse.py [hunk 5]
```diff
@@ -104,76 +165,195 @@ def convert_train_data(self, text_item, symmetry):
...
+                # Downsample to 1/Nth of the items.
+                items = random.sample(items, len(items) // self.sample)
+            return items
+        else:
+            #print("V1 chunkdata")
+            file_chunkdata = chunkdata.splitlines()
+
+            result = []
+            for i in range(0, len(file_chunkdata), DATA_ITEM_LINES):
+                if self.sample > 0:
```
Comment (sethtroisi) [inline training/tf/parse.py]: `if self.sample > 1`

File: training/tf/parse.py [hunk 6]
```diff
@@ -35,6 +41,10 @@
 # if you change this.
 BATCH_SIZE = 512
 
+# Use a random sample of 1/16th of the input data read. This helps
+# improve the spread of games in the shuffle buffer.
+DOWN_SAMPLE = 8
```
Comment (gcp) [inline training/tf/parse.py]: Comment does not match actual value.

File: training/tf/parse.py [hunk 7]
```diff
@@ -65,28 +104,50 @@ def __init__(self, chunks, workers=None):
         # Build the all-zeros and all-ones flat planes, used for color-to-move.
         self.flat_planes = [ b'\0' * 361, b'\1' * 361 ]
 
+        # set the down-sampling rate
+        self.sample = sample
+
+        # V2 Format (see 'PackedMove' in Training.h)
```
Comment (gcp) [inline training/tf/parse.py]: I think this part of the pull request is missing.

File: training/tf/parse.py (hunk 1 above)
Comment (ywrt) [reply training/tf/parse.py]: Not quite sure where that number came from? The minimum size is 2200 (everything zero). The max size depends on the number of non-zero probabilities. The average I measured was about 2500 bytes per move?

File: training/tf/parse.py (hunk 2 above)
Comment (ywrt) [reply training/tf/parse.py]: tnx. fixed.

File: training/tf/parse.py (hunk 5 above)
Comment (ywrt) [reply training/tf/parse.py]: done.

File: training/tf/parse.py (hunk 4 above)
Comment (ywrt) [reply training/tf/parse.py]: done.

File: training/tf/parse.py (hunk 6 above)
Comment (ywrt) [reply training/tf/parse.py]: fixed (me messing with it for testing)

File: training/tf/parse.py (hunk 7 above)
Comment (ywrt) [reply training/tf/parse.py]: I removed the mention. If you're interested in those changes, I'll put them into a separate pull request (that code is still a little messier, and not backward compatible yet).

**Comment by gcp** [comment]
>Looking at the code this should still be backwards compatible with the v1 chuck files? I'm getting the following error:

I have the same problem now. The code isn't compatible with Windows.

Unless a quick fix is possible, this will be backed out.

**Comment by ywrt** [comment]
that's annoying. What version of python is that?

(I'll send through a fix later today)

**Comment by Alderi-Tokori** [comment]
I have the same problem, python 3.5.2

**Pull request merged**